dkg/dkg_test.go

```diff
@@ -0,0 +1,199 @@
+// Copyright © 2022 Obol Labs Inc.
+//
+// This program is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation, either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of  MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package dkg_test
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+	"path"
+	"testing"
+	"time"
+
+	"github.com/coinbase/kryptology/pkg/signatures/bls/bls_sig"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/obolnetwork/charon/app/log"
+	"github.com/obolnetwork/charon/cluster"
+	"github.com/obolnetwork/charon/cmd"
+	"github.com/obolnetwork/charon/dkg"
+	"github.com/obolnetwork/charon/p2p"
+	"github.com/obolnetwork/charon/tbls"
+	"github.com/obolnetwork/charon/testutil"
+	"github.com/obolnetwork/charon/testutil/keystore"
+)
+
+func TestDKG(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	const (
+		nodes = 3
+		vals  = 2
+	)
+
+	// Start bootnode
+	bootnode, errChan := startBootnode(ctx, t)
+
+	// Setup
+	lock, keys, _ := cluster.NewForT(t, vals, 3, 3, 0)
+
+	dir, err := os.MkdirTemp("", "")
+	require.NoError(t, err)
+
+	conf := dkg.Config{
+		DataDir: dir,
+		P2P: p2p.Config{
+			UDPBootnodes: []string{bootnode},
+		},
+		Log:     log.DefaultConfig(),
+		TestDef: &lock.Definition,
+	}
+
+	// Run dkg for each node
+	var eg errgroup.Group
+	for i := 0; i < nodes; i++ {
+		conf := conf
+		conf.DataDir = path.Join(dir, fmt.Sprintf("node%d", i))
+		conf.P2P.TCPAddrs = []string{testutil.AvailableAddr(t).String()}
+		conf.P2P.UDPAddr = testutil.AvailableAddr(t).String()
+
+		require.NoError(t, os.MkdirAll(conf.DataDir, 0o755))
+		err := crypto.SaveECDSA(p2p.KeyPath(conf.DataDir), keys[i])
+		require.NoError(t, err)
+
+		eg.Go(func() error {
+			return dkg.Run(ctx, conf)
+		})
+		if i == 0 {
+			// Allow node0 some time to startup, this just mitigates startup races and backoffs but isn't required.
+			time.Sleep(time.Millisecond * 100)
```

Comment:
@corverroos is there a way to check that a node has started? If not I think we should add it as part of another task

Reply:
There is no way at this point to do that. So this is best effort that allows for fast test to mitigate startup race and backoff.

Reply:
I didn't mean to do it now